providers/aha/handlers.go

```
@@ -0,0 +1,107 @@
+	"github.com/amp-labs/connectors/internal/jsonquery"
+	"github.com/amp-labs/connectors/providers/aha/metadata"
+	"github.com/spyzhov/ajson"
+)
+
+const (
+	pageSizeKey = "per_page"
+	pageSize    = "200"
+	pageKey     = "page"
+	sinceKey    = "created_since"
+	apiVersion  = "v1"
+)
+
+func (c *Connector) buildReadRequest(ctx context.Context, params common.ReadParams) (*http.Request, error) {
+	if params.NextPage != "" {
+		return http.NewRequestWithContext(ctx, http.MethodGet, params.NextPage.String(), nil)
+	}
+
+	path, err := metadata.Schemas.LookupURLPath(c.Module(), params.ObjectName)
+	if err != nil {
+		return nil, err
+	}
+
+	url, err := urlbuilder.New(c.ProviderInfo().BaseURL, apiVersion, path)
+	if err != nil {
+		return nil, err
+	}
+
+	addQueryParams(url, params, 1)
+
+	return http.NewRequestWithContext(ctx, http.MethodGet, url.String(), nil)
+}
+
+func (c *Connector) parseReadResponse(
+	ctx context.Context,
+	params common.ReadParams,
+	request *http.Request,
+	response *common.JSONHTTPResponse,
+) (*common.ReadResult, error) {
+	responseKey := metadata.Schemas.LookupArrayFieldName(c.Module(), params.ObjectName)
+
+	path, err := metadata.Schemas.LookupURLPath(c.Module(), params.ObjectName)
+	if err != nil {
+		return nil, err
+	}
+
+	baseURL, err := urlbuilder.New(c.ProviderInfo().BaseURL, apiVersion, path)
+	if err != nil {
+		return nil, err
+	}
+
+	return common.ParseResult(
+		response,
+		common.GetRecordsUnderJSONPath(responseKey),
+		makeNextRecordsURL(baseURL, params),
+		common.GetMarshaledData,
+		params.Fields,
+	)
+}
+
+func makeNextRecordsURL(baseURL *urlbuilder.URL, params common.ReadParams) common.NextPageFunc {
+	return func(node *ajson.Node) (string, error) {
+		pagination, err := jsonquery.New(node).ObjectRequired("pagination")
+		if err != nil {
+			return "", nil //nolint:nilerr
+		}
+
+		totalPage, err := jsonquery.New(pagination).IntegerRequired("total_pages")
+		if err != nil {
+			return "", err
```

I think similiar to the way you handle ObjectRequired you silently return no next page token. I think the same applies for `total_pages` and `current_page`.

`Required` methods would return error in 2 cases:
* Key not found - in our case this is fine to hanle as empty next page.
* Value type mismatch - you expected integer but for some reason it returns a string. If that would ever happen (unlikely), this error would indicate invalid implementation.

Of course one could argue that key not found also means provider API changed and implementation didn't adapt.

---

If there’s a pagination object, it would definitely have `total_pages` and `current_pages`. That’s why I didn’t feel the need to add it, but doing same for `total_pages` and `current_pages` is also a good thing. i'll update it.

---

Some APIs have different pagination format. Which is crazy. The way you enforced the format works as well. We would need to run into the issue of provider changing pagination contract and how to handle it. Being very strict could be of benefit for majority of connectors.